🌱 Proposal for dynamic informer cache

maxsmythe opened the pull request:
This PR shows how Gatekeeper has forked controller runtime to support the dynamic addition/removal of informers.

Happy to flesh this out if people are interested. Not sure what the correct licensing actions are for moving code across CNCF projects.

This is related to #2159 and #1884

Basically, something like this PR will be necessary to clean up informers once no more controllers are using them.

In the interim, having this code would be helpful to Gatekeeper by eliminating our need to maintain a fork, which has been fairly labor intensive. It also may give other members of the community a way to meet their needs for dynamic watches while waiting for the dynamic controller/reference counting approach to be implemented.

<!-- please add an icon to the title of this PR (see VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠ (:warning:, major), ✨ (:sparkles:, minor), 🐛 (:bug:, patch), 📖 (:book:, docs), or 🌱 (:seedling:, other) -->

<!-- What does this do, and why do we need it? -->

ritazh commented:
@vincepri can you please help take a look at this?

braghettos commented:
We have also a use case where a feature like this one could be really helpful: we are going to generate at runtime CRDs and we would like to watch all the CRs based on the created CRDs.

vincepri commented:
In the queue, should be done reviewing by EOW tomorrow

vincepri commented:
@maxsmythe We should probably add some tests before merging as well

vincepri commented on pkg/cache/informer_cache.go:
```diff
@@ -152,6 +153,24 @@ func (ic *informerCache) GetInformer(ctx context.Context, obj client.Object) (In
 	return i.Informer, err
 }
 
+// GetInformerNonBlocking returns the informer for the obj without waiting for its cache to sync.
+func (ic *informerCache) GetInformerNonBlocking(ctx context.Context, obj client.Object) (Informer, error) {
```
I might be missing something, where is the new code going to be used and how?

vincepri replied:
Got it, thanks for the example. I think overall the approach seems good, there might be more feedback in terms of API changes and whatnot.

I assume to actually use these methods we'll require changes to the exposed interface?

For the non-blocking scenario, could we opt into an option instead of adding a new method?

vincepri replied:
For the controller runtime project specifically, it might be worth discussing how we'll use these facilities to shutdown informers at runtime

cc @alvaroaleman

maxsmythe replied:
Totally cool with adding an option. Are you thinking of:

* adding functional options (e.g. `GetInformer(ctx, obj, Blocking(false), SomeOtherOption())`)
* Adding a straight boolean: `GetInformer(ctx, obj, false)`
* An options object: `GetInformer(ctx, obj, Options{Blocking: false})`

?

vincepri replied:
Probably the first option would be more in line with what we have elsewhere, I was thinking something pretty simple like:

```golang
// Informers knows how to create or fetch informers for different
// group-version-kinds, and add indices to those informers.  It's safe to call
// GetInformer from multiple threads.
type Informers interface {
	// GetInformer fetches or constructs an informer for the given object that corresponds to a single
	// API kind and resource.
	GetInformer(ctx context.Context, obj client.Object, opts ...InformerOption) (Informer, error)

	// GetInformerForKind is similar to GetInformer, except that it takes a group-version-kind, instead
	// of the underlying object.
	GetInformerForKind(ctx context.Context, gvk schema.GroupVersionKind, opts ...InformerOption) (Informer, error)
[...]
}
```

And ultimately define the option such as `NonBlocking()` or similar

vincepri replied:
The above is a breaking change, but it provides callers with a backward compatible approach, and allows us to extend the option further in the future if needed

maxsmythe replied:
Functional options it is!

inteon replied:
@vincepri @maxsmythe I don't think the "Blocking" option will be compatible with reference counting the number of calls to AddEventHandler and RemoveEventHandler. If no calls to AddEventHandler have been made yet, we won't start the informer yet. Only once we have one or more `ResourceEventHandlerRegistration`s, we start the informer. So, I would instead opt for removing the `Blocking` option.
I think DynamicSharedInformerFactory took the same approach (since no ctx is passed to ForResource (see https://github.com/kubernetes/kubernetes/blob/589d6f388680b29f7f7961bda4f4262733b54210/staging/src/k8s.io/client-go/dynamic/dynamicinformer/interface.go#L32).

maxsmythe replied:
@inteon blocking is actually the current behavior. This option allows us to get an informer without blocking-for-sync. As such, I think adding this option is moving things in the direction you're suggesting.

blocking is still necessary for supporting `client.Get()`, which is synchronous and assumes the informer cache is ready to be read from.

Does that allay your concern?

maxsmythe replied:
closing this, since non-blocking get has been moved to #2371

vincepri commented on pkg/cache/internal/informers.go:
```diff
@@ -558,3 +592,50 @@ func restrictNamespaceBySelector(namespaceOpt string, s Selector) string {
 	return ""
 }
+
+// eitherChan returns a channel that is closed when either of the input channels are signaled.
+// The caller must call the returned CancelFunc to ensure no resources are leaked.
+func eitherChan(a, b <-chan struct{}) (<-chan struct{}, context.CancelFunc) {
```
These utilities might be useful elsewhere as well, could we:
- Move them under a new package like `pkg/internal/syncs`
- Use generics so we can reuse for other types of channels as well
- Add some tests

maxsmythe replied:
Absolutely!

vincepri commented on pkg/cache/internal/informers.go:
```diff
@@ -558,3 +592,50 @@ func restrictNamespaceBySelector(namespaceOpt string, s Selector) string {
+	return out, cancelFunc
+}
+
+// mergeChan returns a channel that is closed when any of the input channels are signaled.
+// The caller must call the returned CancelFunc to ensure no resources are leaked.
+func mergeChan(a, b, c <-chan struct{}) (<-chan struct{}, context.CancelFunc) {
```
Same as above

maxsmythe replied:
moved

maxsmythe commented:
Thanks for the review!

Regarding tests, I was holding off until I got signal that there was interest in the idea. It sounds like that hurdle is cleared?

maxsmythe commented on pkg/cache/informer_cache.go:
```diff
@@ -152,6 +153,24 @@ func (ic *informerCache) GetInformer(ctx context.Context, obj client.Object) (In
+		return nil, err
+	}
+
+	// Use a canceled context to signal non-blocking
+	canceledCtx, cancel := context.WithCancel(context.Background())
+	cancel()
```
+1, waiting for feedback on the shape that should take in a different comment:

https://github.com/kubernetes-sigs/controller-runtime/pull/2285#discussion_r1186564862

maxsmythe replied:
Closing this, since non-blocking was moved to #2371

maxsmythe commented:
@vincepri Any feedback on the correct thing to do vis-a-vis licensing, since this code is copied from the Gatekeeper project (which is Apache licensed and CNCF-owned)?

Not sure if this requires a NOTICE file, or a citation in the comment or similar.

vincepri commented:
Where is the code copied from? In general, I have no idea how that works, but I'd expect code for this specific feature to go be net-new in controller runtime, especially with the changes proposed above

braghettos commented:
@maxsmythe @vincepri any updates on the proposed PR? This looks so useful for many use cases!

maxsmythe commented:
@braghettos Sorry, I got distracted. Just pushed a first pass responding to feedback now.

@vincepri I took a stab at addressing your comments. LMK what you think. Also let me know what, if any, additional tests you'd like to see.

Tests pass locally, not sure why there is a failure currently.

maxsmythe commented:
/retest

maxsmythe commented:
Unit tests passed on retest. Still haven't been able to get them to fail locally. Rare flake?

braghettos commented:
Any updates on this PR?

stevekuznetsov commented:
Kubernetes upstream added support for dynamic informer lifecycle handling for the shared informer factory in 1.26. Any chances that we could lean on that?

maxsmythe commented:
@stevekuznetsov This is a step towards being able to lean on that.

For context, here is DynamicInformerFactory's interface:

https://github.com/kubernetes/kubernetes/blob/589d6f388680b29f7f7961bda4f4262733b54210/staging/src/k8s.io/client-go/dynamic/dynamicinformer/interface.go#L25-L49

Which returns a GenericInformer:

https://github.com/kubernetes/kubernetes/blob/589d6f388680b29f7f7961bda4f4262733b54210/staging/src/k8s.io/client-go/informers/generic.go#L74-L79

Which returns a SharedIndexInformer:

https://github.com/kubernetes/client-go/blob/2a5f18df73b70cb85c26a3785b06162f3d513cf5/tools/cache/shared_informer.go#L161-L164

Which allows for the dynamic addition/removal of listeners and whose Run() method accepts a channel to signal the watch should be torn down. I believe this is the dynamic lifecycle handling you're talking about? Let me know if not.

What this PR does is provide a mechanism for sending that teardown signal, but takes no stance on how listeners should be added/removed.

The idea is that, eventually, controller-runtime will leverage the lifecycle methods for adding/removing listeners. Once it does, it will need some kind of method for tearing down watches to avoid memory leaks (I believe reference counting was an idea that had been floated). This PR adds the building block for tearing down watches that will eventually be used.

In the interim, users of controller-runtime who need dynamic watches and are willing to handle engaging with dynamic watches directly (Gatekeeper is one of these users) can use controller runtime without forking it while the community finds the best way forward for a more managed approach.

@vincepri LMK if you have any feedback on the PR. I'm blocked on feedback. Also, WRT the license issue, LMK if you'd like me to reach out to CNCF for this. I believe they have people to give clarity for these kinds of questions.

stevekuznetsov commented:
Awesome, sorry I did not dig in deeply enough to catch that. That approach sounds awesome :)

maxsmythe commented:
Glad to add clarity!

vincepri commented:
In queue, need some more time to look through the changes early next week 😄

maxsmythe commented:
/retest-required

sbueringer commented on pkg/internal/syncs/syncs_test.go:
```diff
@@ -0,0 +1,107 @@
+		signal int
+	}{
+		{
+			name:   "single channel",
```
```suggestion
			name:   "single channel, close 0",
```
nit

Same below

maxsmythe replied:
fixed

sbueringer commented:
Overall looks good to me +/- the linter/compile issues

sbueringer commented on pkg/cache/internal/informers.go:
```diff
@@ -300,6 +339,21 @@ func (ip *Informers) Get(ctx context.Context, gvk schema.GroupVersionKind, obj r
 }
 
+// Remove removes an informer entry and stops it if it was running.
+func (ip *Informers) Remove(gvk schema.GroupVersionKind, obj runtime.Object) {
```
Who  can use this func?

maxsmythe replied:
It's intended to be used by whatever is responsible for knowing when an informer is no longer needed.

In the short term, this responsibility is owned by individual users. Longer-term, it sounds like controller-runtime wants to have some kind of solution for this (reference counting has been floated).

I think it's useful to have this function public as:

1) Otherwise, users of controller-runtime will be blocked from using dynamic informers until a solution is agreed on/implemented (original issue, #1884, is over a year old without a definitive path forward, so this is definitely not a trivial task). Without this, they have no way of cleaning up unused state.

2) Having this method public will provide an escape hatch should controller-runtime's chosen model of managing informers be incompatible with a given project.

sbueringer commented on pkg/cache/internal/informers.go:
```diff
@@ -36,6 +36,7 @@ import (
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/cache"
```
I think unit tests on some level would be good to get coverage for the new functionality

maxsmythe replied:
Added unit tests

inteon commented on pkg/internal/syncs/syncs.go:
```diff
@@ -0,0 +1,37 @@
+	cancel := make(chan T)
+	cancelFunc := func() {
+		once.Do(func() {
+			close(cancel)
```
I think we have to wait for out to be canceled here.
Currently `cancel()` can complete without `chan out` being closed.

maxsmythe replied:
Are you saying that you'd like to see `cancelFunc()` block until `out` is definitely closed?

This is different than how context.WithCancel() [works](https://cs.opensource.google/go/go/+/refs/tags/go1.20.4:src/context/context.go;l=471), but I don't see that as proof-positive it shouldn't be done how you're suggesting.

What is the use case for blocking-until-closed? That won't necessarily guarantee that any goroutines that are expected to respond to `out` have done so (e.g. maybe they are running an infinite for-loop, occasionally checking to see if `out` is closed). To guarantee that, each goroutine would need to send a `done` signal, confirming they have shut down.

Or do you see a bug where `out` may not successfully be closed?

maxsmythe replied:
I'm not sure what an example of code that relies on cancel blocking on channel close would look like? Generally cancel() signals child goroutines that they should exit.

To avoid arguing, I've added it in. It does mean that future coding errors have a higher chance of causing deadlocks, but I think is fine as-written.

inteon commented on pkg/cache/internal/informers.go:
```diff
@@ -92,6 +107,25 @@ type tracker struct {
+type GetOptions struct {
+	// Async instructs Get() to return the informer immediately,
+	// without waiting for its cache to sync.
+	Async bool
```
```suggestion
	WaitForInformerSynced bool
```
Its not very clear to me what `Async` means, maybe its just me.

maxsmythe replied:
This is moved to https://github.com/kubernetes-sigs/controller-runtime/pull/2371

I'll add the suggestion in there.

maxsmythe replied:
Wait... Async was a separate suggestion. Leaving `DoNotBlockUntilSynced` for now, we can debate naming on the new PR.

One thing to note is that `false` should mean "block until ready" in order to maintain backwards compatible behavior when no args are applied.

maxsmythe replied:
Closing this, since non-blocking was moved to #2371

sbueringer commented on pkg/cache/internal/informers.go:
```diff
@@ -92,6 +107,25 @@ type tracker struct {
 
+// GetOptions provides configuration to customize the behavior when
+// getting an informer.
+type GetOptions struct {
```
(Just trying to understand how it fits together)

How is this option related to allowing to stop informers?

Is it that in your scenario the informers are started without waiting for the cache and if they don't come up you call the stop func?

maxsmythe replied:
This bit is more about the utility of and usage patterns for stoppable informers.

In the original code, when a user requests an informer, controller-runtime blocks until that informer's cache is populated. This makes sense for imperative requests, like retrieving a specific object, since the cache needs to be read from for the request to be satisfied. This is not strictly necessary for registering a listener with an informer, since that listener will simply be dormant until events are fed to it.

Gatekeeper has a two-controller model for dynamic informers. There is the main controller and the sub controller. The main controller knows how to derive what GVKs the sub controller should be receiving events for. It then registers a listener with that GVK's associated informer (constructing it, if necessary). This listener passes events to the sub controller's work queue.

I should note I'm being a bit hand-wavy here as Gatekeeper actually uses a centralized structure to keep track of "interested" controllers (essentially a home-grown version of the reference counting solution controller-runtime is considering). We call it the "watch manager".

By allowing this centralized reference counter the ability to get the informer without blocking on it being ready, the watch manager can add/remove listeners without waiting for the informer to bootstrap (it's preferable to have initializing a watch be a synchronous operation so that the managing controller's reconciler can return whether or not a retry is warranted). This reduces the amount of time that main controllers can block each other.

Of course, sub controllers will eventually need to run a `Get()` request in order to run their reconcile loop, and that will be blocked on bootstrapping but:

1) if the sub controller is receiving events, that means the initial `List()` operation of the informer has been completed and therefore bootstrapping is likely also complete, or nearly complete.

2) there are fewer goroutines involved. A blocked sub controller only blocks itself. Main controllers can block each other.

TL;DR: for efficiency's sake, it'd be nice if adding/removing listeners didn't block on a bootstrapping process which they are not dependent on.

alvaroaleman replied:
That makes sense but seems unrelated to the rest of this PR, can this be it's own change? We seem to be unclear about what the best name is for this but other than that this bit seems uncontroversial

maxsmythe replied:
I have no objections to splitting non-blocking-get into its own PR. I'll wait a day or so to open up a new PR to give people time to object.

maxsmythe replied:
Split into https://github.com/kubernetes-sigs/controller-runtime/pull/2371

braghettos commented:
Hi everyone! Any possibility that this issue moves forward?

maxsmythe commented:
Hey, sorry, I got distracted by integrating the latest release of controller-runtime into Gatekeeper. I hope to re-pick-up this thread this week.

maxsmythe commented:
Split the get options into a separate PR. Tests will have to wait until next week, sadly.

maxsmythe commented:
ping

maxsmythe commented:
Waiting for feedback.

sozercan commented:
@sbueringer @vincepri can you ptal again when you get a chance please?

vincepri commented:
changes generally lgtm, can you rebase?

maxsmythe commented:
@vincepri Thanks for looking! Rebased!

Any concerns WRT attribution, since this code is adapted from Gatekeeper's fork? At this point, the only un-modified code is living in `internal/informers.go`.

The `MergeChans` function is IMO new, since it has been reworked to be entirely generic (previously had hardcoded arity and types). Here is a link to the old Gatekeeper code for reference: https://github.com/open-policy-agent/gatekeeper/blob/694e771143fdbefb1630609eea9443e2c0ca7742/third_party/sigs.k8s.io/controller-runtime/pkg/dynamiccache/internal/informers_map.go#L427-L472

LMK if you want me to see if there is anyone we should be asking for advice.

maxsmythe commented:
@vincepri Any updates? Barring licensing questions, code should be good to merge.

maxsmythe commented:
@vincepri ping

vincepri commented:
On my queue for Monday, apologies I’ve just got back

On Fri, Oct 13, 2023 at 3:45 PM Max Smythe ***@***.***> wrote:

> @vincepri <https://github.com/vincepri> ping
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/controller-runtime/pull/2285#issuecomment-1762329677>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXZJ7ZUUSA6WMVSW3PFCQTX7G77LANCNFSM6AAAAAAXHQNPCQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

maxsmythe commented:
No worries!

vincepri commented:
@maxsmythe Chatting with @alvaroaleman it'd be good going forward to have a small design that talks on how we can get to the larger state of dynamic watchers.

alvaroaleman commented:
@maxsmythe given that we do not have anything to dynamically manage watches on controllers, would you be ok if we change this to `StopInformer` but not remove it from the map? It seems that would still solve your problem but not result in confusing behavior if the informer is then later re-created (existing event sources do not get event)

maxsmythe commented:
Thanks for merging!!

@vincepri @alvaroaleman Definitely interested in making this more user-friendly. Lemme know how I can help.

@alvaroaleman The one concern I would have is memory leaks. The main reason to remove the informer is to also remove the cache data, cleaning stale state.

Is there some way to remove the cache but keep a record of interested sources (this could still be a memory leak, but a smaller one)? Is there a way to have two methods (stop informer and remove informer)? Maybe figuring out the edge cases is part of the larger "dynamic watcher" story, that this is more a stopgap on the way to?

alvaroaleman commented:
@maxsmythe memory leak because there is still data in the informer (the state of the world before it was stopped) or because of the informer itself?

alvaroaleman commented:
> Mostly concerned about the state of the world, as that's the larger dataset (there can be a LOT of very large config maps in a cluster, for instance).

Yeah, maybe there is a way to clean up the store in the informer?

> AFAIK the first attempt is https://github.com/kubernetes-sigs/controller-runtime/pull/2159, which appears to be focused on starting/stopping controllers ad-hoc. Is there any other state I should know about? I can sketch some rough thoughts on a Google doc to start with, unless you have a more preferred way of handling these things?

Only the linked issues to my knowledge. And starting to sketch something would be really helpful :) The PRs make it too easy to miss the forest from all the trees because these changes are big.

maxsmythe commented:
Still working on the stories. @alvaroaleman Is there a timeline yet for when a release may be cut that has this?
